<a href="https://colab.research.google.com/github/campospluiza/procesamientoDelHabla/blob/main/Campos_Pereira_Luiza_ULTIMO_DESAFIO.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ULTIMO DESAFIO

**ALUMNA:** Luiza Campos Pereira

**TEMA:** DETECCIÓN DE ANOMALÍAS

# Introducción:

Para este desafío, vamos a desarrollar un modelo que nos permita identificar si un fragmento de texto pertenece al "Martín Fierro" de José Hernández, basándonos en embeddings y utilizando la distancia coseno. Vamos a detallar cada paso del proceso con el código necesario y explicaciones. Siguiendo la consigna, vamos a utilizar el "Fausto" de Estanislao del Campo como texto de comparación.


# Preprocesamiento del dataset:

Ya lo hicimos en un desafío anterior, por lo que copio directamente el codigo utilizado anteriormente

In [2]:
!pip install pdfplumber

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.0/42.0 kB 1.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.5/48.5 kB 3.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 59.2/59.2 kB 4.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.6/5.6 MB 59.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.8/2.8 MB 64.9 MB/s eta 0:00:00


In [3]:
import requests
import os
import pdfplumber

In [4]:
#creo la variable url
url = "https://www.argentina.gob.ar/sites/default/files/hernandez_jose_-_el_gaucho_martin_fierro.pdf"

#hago el get y guardo la respuesta
response = requests.get(url)

#si la respuesta es ok, abro el archivo, descargo los datos y con eso escribo el archivo
if response.status_code == 200:
  file_name = "el_gaucho_martin_fierro.pdf"
  with open(file_name, "wb") as file:
    file.write(response.content)
#imprimo el peso para verificar que se haya descargado correctamente
  file_size = os.path.getsize(file_name)
  print(f"PDF descargado exitosamente, tamaño del archivo: {file_size} bytes")
else:
#veo cual es el código de estado del response si la respuesta no fue 200
  print(f"Error al descargar el PDF. Código de estado: {response.status_code}")

PDF descargado exitosamente, tamaño del archivo: 513796 bytes


In [5]:
# Función para determinar si una página a la que se extraerá el texto
def pagina_a_procesar(texto_pagina):
    # Palabras clave para omitir ciertas secciones como índice, prólogo, anexos, etc.
    palabras_omitir = ['Índice', 'Prólogo', 'Anexo', 'Imágenes', 'Agradecimientos', 'Dedicatoria', 'http',]

    # Revisar si la página contiene alguna palabra que indique que se excluirá
    for palabra in palabras_omitir:
        if palabra.lower() in texto_pagina.lower():
            return False
    return True

# Función para extraer el texto importante del PDF
def extraer_texto(filename):
    texto_relevante = ''

    with pdfplumber.open(filename) as pdf:
        pagina_inicio=0
        pagina_fin=len(pdf.pages)
        # Iterar sobre las páginas en el rango seleccionado
        for pagina_numero in range(pagina_inicio, pagina_fin):
            pagina = pdf.pages[pagina_numero]
            texto_pagina = pagina.extract_text()

            # Si la página no tiene texto o es irrelevante, omitirla
            if texto_pagina and pagina_a_procesar(texto_pagina):
                texto_relevante += texto_pagina + '\n'

    return texto_relevante


 # Extraer el texto relevante del rango de páginas
texto = extraer_texto(file_name)

# Mostrar el texto extraído
print("Texto relevante extraído:")
print(texto)

Texto relevante extraído:
El gaucho
Martín Fierro
José Hernández
Recursos de dominio público
El Gaucho Martín Fierro
I
Aquí me pongo a cantar
al compás de la vigüela,
que el hombre que lo desvela
una pena estrordinaria,
como la ave solitaria
con el cantar se consuela.
Pido a los santos del cielo
que ayuden mi pensamiento:
les pido en este momento
que voy a cantar mi historia
me refresquen la memoria
y aclaren mi entendimiento.
Vengan santos milagrosos,
vengan todos en mi ayuda,
que la lengua se me añuda
y se me turba la vista;
pido a mi Dios que me asista
en una ocasión tan ruda.
Yo he visto muchos cantores,
con famas bien otenidas
y que después de alquiridas
no las quieren sustentar:
parece que sin largar
se cansaron en partidas.
Mas ande otro criollo pasa
Martín Fierro ha de pasar;
nada lo hace recular,
ni las fantasmas lo espantan,
y dende que todos cantan
yo también quiero cantar.
Cantando me he de morir,
cantando me han de enterrar,
y cantando he de llegar
al pie del Eterno Padre;

# Ahora pasamos a procesar el texto FAUSTO que vamos a utilizar para comparar

In [16]:
# Cargar el archivo PDF descargado manualmente
file_name_fausto = "/content/Fausto_-_Estanislao_del_Campo.pdf"
texto_fausto = extraer_texto(file_name_fausto)

In [17]:
# Verificar el texto extraído
print("Texto relevante extraído de Fausto:")
print(texto_fausto[:1000])  # Mostrar solo los primeros 1000 caracteres para revisar

Texto relevante extraído de Fausto:
FAUSTO.
e o
I.M P R E SI ON E S DE L GA U H
, ,
EN LA REPRESENTACION DE ESTA ÓPERA .
•
ESClIITAS POR
ESTANISLAO DEL CAMPO.
PUBLICACION HECH.A
EN FAVOR DE LOS HOSP TALES' MILITARES
BUENOS AIRES
Imp. uBuenos:Aires», Calle de Moreno, frente á,la casa de G,obierno.
~866. ,..
IMPRESIONES DEL GAUCHO
EN LA REPRESENTACIONDE ESTA ÓPERA.
ESCRITAS POR
, ESTANISLAO DEL CAMPO .
.... ").
PUBLlCACION 11 ",".\
EN FAVOR DE LOS HOSPITALES MILITARES.
BUENOS AIHES
,
1m, ••B uenos Aires ... Calle de Moreno, frente á la ca!la de Go)if!ru~.
1866.
Del Ca~po:
Las buenas oLras son siempre hijas de los bellos sentimientos, porque las
mejores y mas grandes ideas nacen en el corazon, llevando consigo la ema
cion de que nacieron.
Su pobreza de poeta, empeñada en aliviar dolol'osos infortunios, ha
apelado á esa infatigable alquimista de la im&jinacion, que elabora los su¿
ñOI de oro y fabrica los palacios en el aire, y ella, evocándole al Demonio,
ha tenido el poder de ponerlo al 

In [18]:
# Crear el corpus de estrofas para "Fausto"
def crear_corpus(texto):
    texto_modificado = re.sub(r'([.!])', r'\1  ', texto)
    estrofas = texto_modificado.split('  ')
    estrofas = [estrofa.strip() for estrofa in estrofas if estrofa.strip()]
    return estrofas

estrofas_fausto = crear_corpus(texto_fausto)

# Criterio del Uso del Dataset

Vamos a dividir el texto de "Martín Fierro" en estrofas para nuestro análisis. Cada estrofa será tratada como una unidad de entrada para el modelo. Usaremos aproximadamente el 80% de las estrofas para el entrenamiento del modelo, y el 20% restante para la validación y prueba. El texto de "Fausto" lo utilizaremos como conjunto de prueba para verificar si el modelo puede distinguirlo del texto de "Martín Fierro".

# Vectorización del Dataset

Para este proyecto, usaremos la técnica de Word Embeddings, más específicamente, un modelo de Word2Vec entrenado sobre el texto de "Martín Fierro". La elección de Word2Vec permite captar las relaciones semánticas y de contexto dentro del texto, lo cual es adecuado para analizar poesía. Luego, utilizaremos la distancia coseno para comparar embeddings de diferentes estrofas y evaluar la similitud.



In [8]:
# Importamos las librerías necesarias
from sklearn.model_selection import train_test_split
from gensim.models import Word2Vec
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np
import re

In [19]:
# Función para crear corpus de estrofas a partir del texto
def crear_corpus(texto):
    # Reemplazar puntos finales y signos de exclamación por doble espacio
    texto_modificado = re.sub(r'([.!])', r'\1  ', texto)

    # Dividir el texto en estrofas usando doble espacio como delimitador
    estrofas = texto_modificado.split('  ')

    # Limpiar estrofas vacías y eliminar espacios adicionales
    estrofas = [estrofa.strip() for estrofa in estrofas if estrofa.strip()]

    return estrofas

# Crear el corpus de estrofas de "Martín Fierro"
corpus_documentos = crear_corpus(texto)

# Dividir el corpus en conjunto de entrenamiento y prueba (80-20)
estrofas_train, estrofas_test = train_test_split(corpus_documentos, test_size=0.2, random_state=42)

# Creamos el corpus de estrofas para "Fausto"
estrofas_fausto = crear_corpus(texto_fausto)

# Entrenamiento del Modelo de Embeddings
Ahora, entrenaremos el modelo de embeddings usando las estrofas del conjunto de entrenamiento. Cada estrofa será tokenizada y usada para entrenar el modelo de Word2Vec.

In [20]:
# Tokenizamos el texto para el modelo Word2Vec
def tokenizar_texto(estrofas):
    return [estrofa.split() for estrofa in estrofas]

# Tokenizamos el conjunto de entrenamiento
tokenized_train = tokenizar_texto(estrofas_train)

# Entrenamos el modelo Word2Vec sobre el conjunto de entrenamiento
modelo_w2v = Word2Vec(sentences=tokenized_train, vector_size=100, window=5, min_count=1, workers=4)


# Función para Obtener el Embedding de una Estrofa
Para obtener el embedding de una estrofa, tomaremos el promedio de los embeddings de cada palabra en la estrofa. Esta estrategia ayuda a obtener una representación general de la estrofa completa.

In [21]:
# Función para obtener el embedding promedio de una estrofa
def obtener_embedding_estrofa(estrofa, modelo):
    tokens = estrofa.split()
    embedding = np.mean([modelo.wv[token] for token in tokens if token in modelo.wv], axis=0)
    return embedding

# Cálculo de la Similaridad Coseno
Utilizaremos la distancia coseno entre el embedding de una estrofa y los embeddings del conjunto de entrenamiento para verificar si pertenece al "Martín Fierro".

In [22]:
# Función para calcular la similitud coseno entre estrofas
def calcular_similitud(estrofa, modelo, embeddings_train):
    embedding_estrofa = obtener_embedding_estrofa(estrofa, modelo).reshape(1, -1)
    similitudes = cosine_similarity(embedding_estrofa, embeddings_train)
    return np.mean(similitudes)

# Embeddings del Conjunto de Entrenamiento y Prueba
Para evaluar el rendimiento del modelo, primero calculamos los embeddings del conjunto de entrenamiento, y luego evaluaremos con el conjunto de prueba de "Martín Fierro" y con "Fausto".

In [25]:
# Función para obtener el embedding promedio de una estrofa
def obtener_embedding_estrofa(estrofa, modelo):
    tokens = estrofa.split()
    # Filtrar solo las palabras que están en el vocabulario del modelo
    palabras_en_vocabulario = [modelo.wv[token] for token in tokens if token in modelo.wv]

    # Si no hay palabras en el vocabulario, devolver un vector de ceros
    if len(palabras_en_vocabulario) == 0:
        return np.zeros(modelo.vector_size)

    # Calcular el embedding promedio
    embedding = np.mean(palabras_en_vocabulario, axis=0)
    return embedding

# Recalculamos los embeddings del conjunto de entrenamiento
embeddings_train = np.array([obtener_embedding_estrofa(estrofa, modelo_w2v) for estrofa in estrofas_train])

# Evaluar el conjunto de prueba de "Martín Fierro" y "Fausto"
# Filtrar solo las estrofas con embeddings válidos (no vacíos) para similitudes
similaridades_mf_test = [calcular_similitud(estrofa, modelo_w2v, embeddings_train)
                         for estrofa in estrofas_test if np.any(obtener_embedding_estrofa(estrofa, modelo_w2v))]
similaridades_fausto = [calcular_similitud(estrofa, modelo_w2v, embeddings_train)
                        for estrofa in estrofas_fausto if np.any(obtener_embedding_estrofa(estrofa, modelo_w2v))]

# Mostrar resultados
print("Similitud promedio para estrofas de Martín Fierro (test):", np.mean(similaridades_mf_test))
print("Similitud promedio para estrofas de Fausto:", np.mean(similaridades_fausto))


Similitud promedio para estrofas de Martín Fierro (test): 0.7195095
Similitud promedio para estrofas de Fausto: 0.55566436


# Análisis de Resultados
En este proyecto, el objetivo era construir un modelo capaz de detectar si un fragmento de texto proviene del poema gauchesco "Martín Fierro" de José Hernández o de un texto distinto, en este caso, el poema "Fausto" de Estanislao del Campo. Para lograr esto, seguimos los siguientes pasos:


*   Preparación y Tokenización del Texto: Se descargaron ambos textos en formato PDF, extrayendo y limpiando el contenido relevante, y luego se dividieron en estrofas para ser procesadas como unidades individuales de texto.
*   Entrenamiento del Modelo de Embeddings: Utilizamos el texto de "Martín Fierro" para entrenar un modelo de embeddings (Word2Vec), que capturó las relaciones semánticas y de contexto entre las palabras del poema.
*   Similitud Coseno: Después de obtener embeddings promedio de cada estrofa, calculamos la similitud coseno entre las estrofas del conjunto de prueba y los embeddings del conjunto de entrenamiento de "Martín Fierro". También calculamos las similitudes entre el conjunto de estrofas de "Fausto" y el conjunto de entrenamiento de "Martín Fierro". Esto permitió evaluar qué tan cercanas eran las estrofas de "Fausto" y de "Martín Fierro" respecto a los patrones semánticos capturados por el modelo.

# Resultados Obtenidos
Similitud promedio para estrofas de "Martín Fierro" (conjunto de prueba): 0.7195

Similitud promedio para estrofas de "Fausto": 0.5557

Estos resultados indican una diferencia notable en la similitud promedio entre las estrofas de "Martín Fierro" y las de "Fausto" cuando se comparan con el modelo de embeddings entrenado en "Martín Fierro".

# Interpretación

La similitud promedio de 0.7195 para las estrofas de "Martín Fierro" en el conjunto de prueba sugiere que el modelo de embeddings captura adecuadamente el estilo y las características lingüísticas del poema. Esta similitud relativamente alta indica que las estrofas de "Martín Fierro" comparten un contexto semántico consistente, que el modelo es capaz de reconocer.

Por otro lado, la similitud promedio de 0.5557 para las estrofas de "Fausto" muestra que, aunque ambos poemas son de estilo gauchesco y comparten algunas características temáticas, existen diferencias semánticas y de estilo que el modelo percibe. Esta diferencia de 0.1638 puntos en la similitud promedio confirma que el modelo puede distinguir entre estrofas del "Martín Fierro" y de "Fausto".

# Conclusiones
Este análisis demuestra que el modelo de embeddings basado en Word2Vec es una herramienta efectiva para capturar el estilo distintivo de "Martín Fierro" y diferenciarlo de otro poema gauchesco como "Fausto". La técnica de similitud coseno permite establecer un umbral de diferencia, donde los valores más altos de similitud están asociados a las estrofas que realmente pertenecen al "Martín Fierro".

Estos resultados sugieren que el enfoque de embeddings y distancia coseno es adecuado para detectar anomalías o diferencias en textos de poesía, incluso cuando los textos comparten el mismo género literario. En futuras implementaciones, se podrían explorar otros modelos de embeddings o técnicas basadas en transformers para capturar aún mejor los matices del lenguaje y mejorar la capacidad de clasificación.

# Mi Experiencia y el uso de ChatGPT
Sobre mi experiencia trabajando este desafio, me pareció bastante dificil. Primero, me costó encontrar una version adecuada del poema Fausto. Primero porque no encontraba el autor como Estanislao Lopez, pero si del Campo. Después porque la mayoría eran pdf de imagenes, donde no se captaba bien el texto o eran url que no me permitian el download. Lo solucioné descargando una versión que estaba medianamente buena y cargando a partir de este archivo y no de la URL.

Después necesité la ayuda de chatgpt, pues con mi primera solución, algunos embeddings generaban un valor NaN. Expliqué a la IA que estaba haciendo y el error y me tiró la solución ajustando el código para 1, revisar si el embedding de una estrofa era vacío. Modificó la función "obtener_embedding_estrofa" para verificar que al menos una palabra de la estrofa estuviera en el vocabulario del modelo antes de calcular el promedio. Si ninguna palabra estaba en el vocabulario, el codigo devuelve un vector de ceros en lugar de NaN.2, ignorar estrofas con embeddings vacíos. En el cálculo de similitud coseno, solo consideró estrofas que tienen embeddings válidos. Con ese ajuste hecho por la IA, pude concluir el trabajo.

# Referencias

https://radimrehurek.com/gensim/auto_examples/tutorials/run_word2vec.html#sphx-glr-auto-examples-tutorials-run-word2vec-py

https://nlp.stanford.edu/IR-book/